sdk/python/lib/pulumi/runtime/invoke.py

```diff
@@ -143,3 +145,114 @@ async def do_rpc():
     if invoke_error is not None:
         raise invoke_error
     return InvokeResult(invoke_result)
+
+
+def call(tok: str, props: 'Inputs', res: Optional['Resource'] = None, typ: Optional[type] = None) -> 'Output[Any]':
+    """
+    call dynamically invokes the function, tok, which is offered by a provider plugin.  The inputs
+    can be a bag of computed values (Ts or Awaitable[T]s).
+    """
+    log.debug(f"Calling function: tok={tok}")
+
+    if typ and not _types.is_output_type(typ):
+        raise TypeError("Expected typ to be decorated with @output_type")
+
+    # Setup the futures for the output.
+    resolve_value: 'asyncio.Future' = asyncio.Future()
+    resolve_is_known: 'asyncio.Future[bool]' = asyncio.Future()
+    resolve_is_secret: 'asyncio.Future[bool]' = asyncio.Future()
+    resolve_deps: 'asyncio.Future[Set[Resource]]' = asyncio.Future()
+
+    from .. import Output  # pylint: disable=import-outside-toplevel
+    out = Output(resolve_deps, resolve_value, resolve_is_known, resolve_is_secret)
+
+    async def do_call():
+        try:
+            # Construct a provider reference from the given provider, if one is available on the resource.
+            provider_ref, version = None, ""
+            if res is not None:
+                if res._provider is not None:
+                    provider_urn = await res._provider.urn.future()
+                    provider_id = (await res._provider.id.future()) or rpc.UNKNOWN
+                    provider_ref = f"{provider_urn}::{provider_id}"
+                    log.debug(f"Call using provider {provider_ref}")
+                version = res._version or ""
```

Comment:
Regular invokes accept an `InvokeOptions` which allows specifying the provider and version.

Having a generated method accept `InvokeOptions` as an argument felt unnatural to me since, if you wanted to specify a custom provider, you would have done so when creating the resource itself.

So I've made it so that the provider / version is picked up from the resource, which is passed into this function as part of the generated SDK code, e.g.:

```python
    def get_kubeconfig(__self__,
                       profile_name: Optional[str] = None,
                       role_arn: Optional[str] = None) -> pulumi.Output['Cluster.GetKubeconfigResult']:
        __args__ = dict()
        __args__['__self__'] = __self__
        __args__['profileName'] = profile_name
        __args__['roleArn'] = role_arn
        return pulumi.runtime.call('eks:index:Cluster/getKubeconfig', __args__, res=__self__, typ=Cluster.GetKubeconfigResult)
```

There are two cases that I need to handle in subsequent changes:

1. If we are going to use `Call` for leave functions that accept inputs / create resources, we will want to support `InvokeOptions` for that, so I'll need to update this function to be able to pass either a resource or `InvokeOptions` to specify the custom provider.

2. If the instance of the resource was rehydrated from a resource reference, I don't think the `res._provider` or `res._version` will be available to us. For that case, we'll need to make sure we can fill in that information from the built-in `pulumi:pulumi:getResource`.

Reply:
> If we are going to use Call for leave functions that accept inputs / create resources, we will want to support InvokeOptions for that, so I'll need to update this function to be able to pass either a resource or InvokeOptions to specify the custom provider.

I believe that we will use Call for this as well, yes. I think that we can probably get away with eliding options from methods, but providing them for module-level functions.